Add preview to Threshold config & option flow

jpbede opened the pull request:
## Proposed change
Add the preview function to the Threshold helper config flow and option flow.

Context: The use of the upper, lower or both values together and their effects on the binary sensor are sometimes somewhat confusing (https://github.com/home-assistant/core/issues/110904) for users. With the preview, a user would see the result of their changes directly.

Frontend PR: https://github.com/home-assistant/frontend/pull/19845

<img src="https://github.com/home-assistant/core/assets/3989428/ad866fc8-eb5b-488c-9b52-a0ead311f1b8" width="450">
<img src="https://github.com/home-assistant/core/assets/3989428/26754219-32af-40f0-91ee-00e2ebd9f1a0" width="450">

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [x] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] I have followed the [perfect PR recommendations][perfect-pr]
- [x] The code has been formatted using Ruff (`ruff format homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [x] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/development_checklist/
[manifest-docs]: https://developers.home-assistant.io/docs/creating_integration_manifest/
[quality-scale]: https://developers.home-assistant.io/docs/integration_quality_scale_index/
[docs-repository]: https://github.com/home-assistant/home-assistant.io
[perfect-pr]: https://developers.home-assistant.io/docs/review-process/#creating-the-perfect-pr

jpbede committed:
Add preview to Threshold config & option flow

jpbede committed:
Add preview to Threshold config & option flow

jpbede commented:
> I think perhaps add a printscreen to the PR description to visualize how it will look 👍

Done 🙂

gjohansson-ST reviewed (changes requested):
Conflicts needs to be addressed

jpbede committed:
Fix import

jpbede committed:
Fix import

emontnemery committed:
Merge branch 'dev' into threshold-config-flow-preview

emontnemery commented on homeassistant/components/threshold/binary_sensor.py:
```diff
@@ -260,6 +283,14 @@ def above(sensor_value: float, threshold: float) -> bool:
             self._state = None
             return
 
+        # guard against the case where the thresholds are not set
+        if not hasattr(self, "_threshold_lower") and not hasattr(
+            self, "_threshold_upper"
+        ):
+            self._state_position = POSITION_UNKNOWN
+            self._state = None
+            return
```
Why is this case added? If it's only needed for preview, it should be possible to handle it in `async_start_preview`, if it's a bug fix it should be moved to a separate PR.

jpbede replied:
Yea, that's just the case for preview. I moved this to `async_start_preview`. See the new PR as I screwed up the rebase

pantonvich committed:
Correct state class of ecowitt hourly rain rate sensors (#110475)

* Update sensor.py for Hourly Rain Rates mm and in

hourly_rain_rate from integration ecowitt has state class total_increasing, but its state is not strictly increasing

* Update sensor.py format

gjohansson-ST committed:
Add `open` state to LockEntity (#111968)

* Add `open` state to LockEntity

* Add tests

* Fixes

* Fix tests

* strings and icons

* Adjust demo open lock

* Fix lock and tests

* fix import

* Fix strings

* mute ruff

* Change sequence

* Sequence2

* Group on states

* Fix ruff

* Fix tests

* Add more test cases

* Sorting

tizianodeg committed:
Fix nibe_heatpump climate for models without cooling support (#114599)

* fix nibe_heatpump climate for models without cooling support

* add test for set temperature with no cooling support

* fixup use self._coil_setpoint_cool None

* fixup add new test to explicitly test unsupported cooling

bdraco committed:
Improve config entry has already been setup error message (#117091)

MatthewFlamm committed:
Fix nws blocking startup (#117094)

Co-authored-by: J. Nick Koston <nick@koston.org>

jbouwh committed:
Make the mqtt discovery update tasks eager and fix race (#117105)

* Fix mqtt discovery race for update rapidly followed on creation

* Revert unrelated renaming local var

bdraco committed:
Migrate baf to use config entry runtime_data (#117081)

jbouwh committed:
Add open and opening state support to MQTT lock (#117110)

bdraco committed:
Avoid storing Bluetooth scanner in hass.data (#117074)

bdraco committed:
Migrate govee_ble to use config entry runtime_data (#117076)

bdraco committed:
Avoid creating a task to clear the hass instance at test teardown (#117103)

bdraco committed:
Avoid creating inner tasks to load storage (#117099)

bdraco committed:
Speed up tests by making mock_get_source_ip session scoped (#117096)

bdraco committed:
Migrate yalexs_ble to use config entry runtime_data (#117082)

bdraco committed:
Migrate elkm1 to use config entry runtime_data (#117077)

Lash-L committed:
Bump python-roborock to 2.1.1 (#117078)

jbouwh committed:
Export MQTT subscription helpers at integration level (#116150)

autinerd committed:
Enable Ruff RUF010 (#115371)

Co-authored-by: J. Nick Koston <nick@koston.org>

autinerd committed:
Ignore Ruff SIM103 (#115732)

Co-authored-by: J. Nick Koston <nick@koston.org>

puddly committed:
Add a missing `addon_name` placeholder to the SkyConnect config flow (#117089)

bdraco committed:
Fix config entry _async_process_on_unload being called for forwarded platforms (#117084)

mib1185 committed:
Store runtime data inside the config entry in Tuya (#116822)

mletenay committed:
Goodwe Increase max value of export limit to 200% (#117090)

Lash-L committed:
Add test data for Zeo and Dyad devices to Roborock (#117054)

mletenay committed:
Bump goodwe to 0.3.5 (#117115)

Lash-L committed:
Add better testing to vacuum platform (#112523)

* Add better testing to vacuum platform

* remove state strings

* some of the MR comments

* move MockVacuum

* remove manifest extra

* fix linting

* fix other linting

* Fix create entity calls

* Format

* remove create_entity

* change to match notify

---------

Co-authored-by: Martin Hjelmare <marhje52@gmail.com>

cdce8p committed:
Use HassKey for importlib helper (#117116)

bdraco committed:
Move available property in BasePassiveBluetoothCoordinator to PassiveBluetoothDataUpdateCoordinator (#117056)

dependabot[bot] committed:
Bump github/codeql-action from 3.25.3 to 3.25.4 (#117127)

joostlek committed:
Add airgradient integration (#114113)

joostlek committed:
Add base entity for Airgradient (#117135)

dieselrabbit committed:
Add screenlogic service tests (#116356)

bdraco committed:
Speed up test teardown when no config entries are loaded (#117095)

Avoid the gather call when there are no loaded config entries

jbouwh committed:
Fix typo in xiaomi_ble translation strings (#117144)

Thomas55555 committed:
Catch auth exception in husqvarna automower (#115365)

* Catch AuthException in Husqvarna Automower

* don't use getattr

* raise ConfigEntryAuthFailed

autinerd committed:
Bump ruff to 0.4.4 (#117154)

bdraco committed:
Handle tilt position being None in HKC (#117141)

tizianodeg committed:
Raise ServiceValidationError in Nibe climate services (#117171)

Fix ClimateService to rise ServiceValidationError for stack free logs

RJPoelstra committed:
Improve base entity state in Vogel's MotionMount integration (#109043)

* Update device info when name changes

* Entities now report themselves as being unavailable when the MotionMount is disconnected

* Don't update device_info when name changes

* Use `device_entry` property to update device name

* Assert device is available

Co-authored-by: Erik Montnemery <erik@montnemery.com>

* Add missing import

---------

Co-authored-by: Erik Montnemery <erik@montnemery.com>

CoRfr committed:
Add standard deviation calculation to group (#112076)

* Add standard deviation calculation to group

* Add missing bits

---------

Co-authored-by: G Johansson <goran.johansson@shiftit.se>

jpbede committed:
Merge branch 'threshold-config-flow-preview' of github.com:jpbede/home-assistant-core into threshold-config-flow-preview

jpbede committed:
Move guard to `async_start_preview`

jpbede commented:
Sry everyone, I screwed the rebase
